### ISSUE TITLE
"Last change by" is not always correct for FUGR includes

### ISSUE THREAD

**Issue by FreHu** [abapGit/abapGit#1428]
All FUGR includes return the same last change by (last user who changed something in the FUGR)

**Comment by jfilak** [comment]
What kind of object?

**Comment by larshp** [comment]
The logic for finding the "last changed by" is located in class ZCL_ABAPGIT_OBJECT_FUGR method ZIF_ABAPGIT_OBJECT~CHANGED_BY, can you help to look up what he/she changed and checking in the debugger?

It is difficult to trace remotely

**Comment by jfilak** [comment]
I believe the problem is that FUGR's changed_by method returns the last user who touched any of the FUGR's includes.

**Comment by larshp** [comment]
And which user does SAP standard report as last user?

**Comment by larshp** [comment]
currently not able to reproduce this issue

**Comment by FreHu** [comment]
I will try to make an easy to reproduce example to see if this persists at some point.

**Comment by FreHu** [comment]
Repro steps:

1) Create a package with a FUGR and some includes as user XXX, version it in abapgit
2) Get user YYY to change and activate any of the includes and not touch anything else
3) The user who last changed the whole FUGR as well as every include is now user YYY

@jfilak is right, the correct information is retrieved here for each include:

```zcl_abapgit_object_fugr->zif_abapgit_object~changed_by```
```abap
LOOP AT lt_includes ASSIGNING <lv_include>.
      SELECT unam AS user udat AS date utime AS time FROM reposrc
        APPENDING CORRESPONDING FIELDS OF TABLE lt_stamps
        WHERE progname = <lv_include>
        AND   r3state = 'A'.                              "#EC CI_SUBRC
    ENDLOOP.
```
![2018-08-16_11-02-46](https://user-images.githubusercontent.com/5097067/44199605-e8c6f080-a144-11e8-8d30-441ddc9d77e2.png)

but then the info is lost when the dates are sorted and the latest is used for the whole group.
```abap
SORT lt_stamps BY date DESCENDING time DESCENDING.

    READ TABLE lt_stamps INDEX 1 ASSIGNING <ls_stamp>.
    IF sy-subrc = 0.
      rv_user = <ls_stamp>-user.
    ELSE.
      rv_user = c_user_unknown.
    ENDIF.
```

**Comment by EduardoCopat** [comment]
I could reproduce the issue...

A possible solution is to change `changed_by` to receive the `filename` as an input parameter. In this way, we could know which function group includes is being processed. See below:

![image](https://user-images.githubusercontent.com/2248490/44548625-f5bf8300-a6f4-11e8-80a8-93b463334174.png)

This would enable us to identify who lastly modified an include.

Also, `<ls_local>` could be passed as a parameter, instead of only `<ls_local>-file`. But this would trigger changes in all "serializable" objects.

Any thoughts on this?

**Comment by larshp** [comment]
last changed by has been removed since this issue was created

**Comment by mbtools** [comment]
"Last Changed" is still used on Diff (ZCL_ABAPGIT_GUI_PAGE_DIFF->RENDER_DIFF_HEAD):

![image](https://user-images.githubusercontent.com/59966492/87590036-29727000-c6b4-11ea-89d2-d032600d0930.png)

For objects with only one file (PROG), the current logic makes sense. For objects with multiple files (FUGR, but also CLAS which can have another file for test classes), ZIF_ABAPGIT_OBJECT~CHANGED_BY is simply not sufficient. 

I like the suggestion of @EduardoCopat. Let's pass the filename to CHANGED_BY as an additional, _optional_ parameter. Then classes for objects with multiple files can be enhanced to return the correct user (without impactive the rest). 

It won't work for methods or other parts of a class which is probably the most common scenario, however.

**Comment by FreHu** [comment]
Although SAP can version methods separately, it doesn't fit in the whole git concept which is file based. File-by-file last changed is important as you can have many includes under a FUGR, especially in older code. It is very messy to figure out who did what when you have more than one developer on the same system.

**Comment by mbtools** [comment]
Closed by #6274
